thrust/thrust/system/cuda/detail/merge.h

```diff
@@ -38,833 +38,158 @@ j * Copyright (c) 2016, NVIDIA CORPORATION.  All rights reserved.
-          typename Derived,
-          typename KeysIt1,
-          typename KeysIt2,
-          typename ItemsIt1,
-          typename ItemsIt2,
-          typename KeysOutputIt,
-          typename ItemsOutputIt,
-          typename CompareOp>
-THRUST_RUNTIME_FUNCTION pair<KeysOutputIt, ItemsOutputIt> merge(
+          class CompareOp = less<>>
+pair<KeysOutputIt, ItemsOutputIt> _CCCL_HOST_DEVICE merge_by_key(
   execution_policy<Derived>& policy,
-  KeysIt1 keys1_first,
-  KeysIt1 keys1_last,
-  KeysIt2 keys2_first,
-  KeysIt2 keys2_last,
-  ItemsIt1 items1_first,
-  ItemsIt2 items2_first,
-  KeysOutputIt keys_result,
-  ItemsOutputIt items_result,
-  CompareOp compare_op)
+  KeysIt1 keys1_begin,
+  KeysIt1 keys1_end,
+  KeysIt2 keys2_begin,
+  KeysIt2 keys2_end,
+  ItemsIt1 items1_begin,
+  ItemsIt2 items2_begin,
+  KeysOutputIt keys_out_begin,
+  ItemsOutputIt items_out_begin,
+  CompareOp compare_op = {})
 {
   using size_type = typename iterator_traits<KeysIt1>::difference_type;
 
-  size_type num_keys1 = static_cast<size_type>(thrust::distance(keys1_first, keys1_last));
-  size_type num_keys2 = static_cast<size_type>(thrust::distance(keys2_first, keys2_last));
-
-  size_type const count = num_keys1 + num_keys2;
-
-  if (count == 0)
+  const auto num_keys1    = static_cast<size_type>(distance(keys1_begin, keys1_end));
+  const auto num_keys2    = static_cast<size_type>(distance(keys2_begin, keys2_end));
+  const auto num_keys_out = num_keys1 + num_keys2;
+  if (num_keys_out == 0)
   {
-    return thrust::make_pair(keys_result, items_result);
+    return {keys_out_begin, items_out_begin};
   }
 
-  size_t storage_size = 0;
-  cudaStream_t stream = cuda_cub::stream(policy);
-
+  const auto stream = cuda_cub::stream(policy);
   cudaError_t status;
-  status = doit_step<MERGE_ITEMS>(
-    nullptr,
-    storage_size,
-    keys1_first,
-    keys2_first,
-    items1_first,
-    items2_first,
+  size_t storage_size = 0;
+  THRUST_DOUBLE_INDEX_TYPE_DISPATCH(
+    status,
+    cub::DeviceMerge::MergePairs,
     num_keys1,
     num_keys2,
-    keys_result,
-    items_result,
-    compare_op,
-    stream);
-  cuda_cub::throw_on_error(status, "merge: failed on 1st step");
-
-  // Allocate temporary storage.
-  thrust::detail::temporary_array<thrust::detail::uint8_t, Derived> tmp(policy, storage_size);
-  void* ptr = static_cast<void*>(tmp.data().get());
-
-  status = doit_step<MERGE_ITEMS>(
-    ptr,
-    storage_size,
-    keys1_first,
-    keys2_first,
-    items1_first,
-    items2_first,
+    (nullptr,
+     storage_size,
+     keys1_begin,
+     items1_begin,
+     num_keys1_fixed,
+     keys2_begin,
+     items2_begin,
+     num_keys2_fixed,
+     keys_out_begin,
+     items_out_begin,
+     compare_op,
+     stream));
+  throw_on_error(status, "merge: failed on 1st step");
+
+  thrust::detail::temporary_array<char, Derived> temp_storage(policy, storage_size);
+  THRUST_DOUBLE_INDEX_TYPE_DISPATCH(
+    status,
+    cub::DeviceMerge::MergePairs,
```

Comment:
important: this macro will lead to invocation of `cub::DeviceMerge::MergePairs`, which accepts `num_keys{1,2}` as `int`. So, both 32 and 64-bit invocations just lead to `int` in the end. For now, the conventional pattern is to use something along the lines of:
```cpp
  using Dispatch32 = cub::DispatchMerge</* ... */ thrust::detail::int32_t>;
  using Dispatch64 = cub::DispatchMerge</* ... */ thrust::detail::int64_t>;
```
and then:
```cpp
  THRUST_DOUBLE_INDEX_TYPE_DISPATCH2(
    status,
    Dispatch32::Dispatch,
    Dispatch64::Dispatch,
    num_items,
    (d_temp_storage, temp_storage_bytes, first, result, num_items_fixed, binary_op, stream));
```

Reply:
Yeah, your previous comments to hardcode the CUB API to `int` broke that :D I will adapt the Thrust call accordingly.